receiver/awsxrayreceiver/internal/translator/translator.go

```diff
@@ -0,0 +1,190 @@
+	spans := ils.Spans()
+
+	// populating global attributes shared among segment and embedded subsegment(s)
+	populateInstrumentationLibrary(&seg, &ils)
+	populateResource(&seg, &resource)
+
+	// recursively traverse segment and embedded subsegments
+	// to populate the spans. We also need to pass in the
+	// TraceID of the root segment in because embedded subsegments
+	// do not have that information, but it's needed after we flatten
+	// the embedded subsegment to generate independent child spans.
+	_, _, err = segToSpans(seg, seg.TraceID, nil, &spans, 0)
+	if err != nil {
+		return nil, count, err
+	}
+
+	return &traceData, count, nil
+}
+
+func segToSpans(seg tracesegment.Segment,
+	traceID, parentID *string,
+	spans *pdata.SpanSlice, startingIndex int) (int, *pdata.Span, error) {
+
+	span := spans.At(startingIndex)
+
+	err := populateSpan(&seg, traceID, parentID, &span)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	startingIndexForSubsegment := 1 + startingIndex
+	var populatedChildSpan *pdata.Span
+	for _, s := range seg.Subsegments {
+		startingIndexForSubsegment, populatedChildSpan, err = segToSpans(s,
+			traceID, seg.ID,
+			spans, startingIndexForSubsegment)
+		if err != nil {
+			return 0, nil, err
+		}
+
+		if seg.Cause != nil &&
+			populatedChildSpan.Status().Code() != pdata.StatusCode(otlptrace.Status_Ok) {
+			// if seg.Cause is not nil, then one of the subsegments must contain a
+			// HTTP error code. ALso, span.Status().Code() is already
+			// set to `otlptrace.Status_UnknownError` by `addCause()` in
+			// `populateSpan()` above, so here we are just trying to figure out
+			// whether we can get a event more specific error code.
+
+			if span.Status().Code() == pdata.StatusCode(otlptrace.Status_UnknownError) {
+				// update the error code to a possibly more specific code
+				span.Status().SetCode(populatedChildSpan.Status().Code())
+			}
+		}
+	}
+
+	return startingIndexForSubsegment, &span, nil
+}
+
+func populateSpan(
+	seg *tracesegment.Segment,
+	traceID, parentID *string,
+	span *pdata.Span) error {
+
+	attrs := span.Attributes()
+	attrs.InitEmptyWithCapacity(initAttrCapacity)
+
+	err := addNameAndNamespace(seg, span)
+	if err != nil {
+		return err
+	}
+
+	if seg.TraceID == nil {
+		// if seg.TraceID is nil, then `seg` must be an embedded subsegment.
+		span.SetTraceID(pdata.TraceID([]byte(*traceID)))
+	} else {
+		span.SetTraceID(pdata.TraceID([]byte(*seg.TraceID)))
+	}
+	span.SetSpanID(pdata.SpanID([]byte(*seg.ID)))
+	addParentSpanID(seg, parentID, span)
+	addStartTime(seg.StartTime, span)
+
+	addEndTime(seg.EndTime, span)
+	addBool(seg.InProgress, AWSXRayInProgressAttribute, &attrs)
+	addString(seg.User, conventions.AttributeEnduserID, &attrs)
+
+	addHTTP(seg, span)
+	addCause(seg, span)
+	addAWSToSpan(seg.AWS, &attrs)
+	err = addSQLToSpan(seg.SQL, &attrs)
+	if err != nil {
+		return err
+	}
+
+	addBool(seg.Traced, AWSXRayTracedAttribute, &attrs)
+
+	// TODO: transform the `metadata` field once X-Ray exporter
```

Comment:
x-ray exporter will deal with metadata differently for OTel instrumented apps than this receiver I think. There is no concept of metadata / annotation in OTel, so we'll use key lists to control which it becomes (/cc @shengxil) since prefixing is complicated and breaks compliance with other backends.

For this receiver, we should instead define a convention prefix, maybe `opentelemetry.internal.collector.<metadata key> = JSON.serialize(<metadata value>)` and process that specially in the exporter.

Reply:
As you mentioned, there's no concept in OT between attributes that will be indexed (i.e. X-Ray annotations) and those that will not be (X-Ray metadata). So for now I will go ahead and define a prefix that will only be used by the X-Ray OT receiver and exporter